worldedit-core/src/main/java/com/sk89q/worldedit/LocalSession.java

```diff
@@ -405,6 +406,15 @@ public void setTimezone(ZoneId timezone) {
      */
     public void clearHistory() {
         //FAWE start
+        if (Fawe.isMainThread() && historyWriteLock.isLocked()) {
```

Comment:
The only "negative" outcome of the race condition would be if the lock is locked whilst waiting for the async task to complete, in which case we just wait in the async task which is fine, right? For the sake of code readability and simplicity I'm fine with it (very extremely rarely) submitting an async task when it could otherwise be handled on the main thread

Reply:
I was thinking of the following scenario:
- we are on the main thread
- isLocked is false
- after the call, another thread acquires the lock
- we call clearHistoryTask
- we have to wait on the main thread until the other thread releases the lock

I don't exactly know in which cases and for what other threads acquire the lock, but that seems similar to the original issue?

Reply:
Eh I suppose so yeah